### PR TITLE
[RS-1278] Desktop tiles forecasting inputs table

### PR DESCRIPTION
**Pull request by sergiosonline**
Checklist for reviewer:

- [ ] Commits should reference a bug or github issue, if relevant (if a bug is referenced, the pull request should include the bug number in the title).
- [ ] If the PR comes from a fork, trigger integration CI tests by running the [Push to upstream workflow](https://github.com/mozilla/bigquery-etl/actions/workflows/push-to-upstream.yml) and provide the `<username>:<branch>` of the fork as parameter. The parameter will also show up
in the logs of the `manual-trigger-required-for-fork` CI task together with more detailed instructions.
- [ ] If adding a new field to a query, ensure that the schema and dependent downstream schemas have been updated.
- [ ] When adding a new derived dataset, ensure that data is not available already (fully or partially) and recommend extending an existing dataset in favor of creating new ones. Data can be available in the [bigquery-etl repository](https://github.com/mozilla/bigquery-etl), [looker-hub](https://github.com/mozilla/looker-hub) or in [looker-spoke-default](https://github.com/mozilla/looker-spoke-default/tree/e1315853507fc1ac6e78d252d53dc8df5f5f322b).

For modifications to schemas in restricted namespaces (see [`CODEOWNERS`](https://github.com/mozilla/bigquery-etl/blob/main/CODEOWNERS)):
- [ ] Follow the [change control procedure](https://docs.google.com/document/d/1TTJi4ht7NuzX6BPG_KTr6omaZg70cEpxe9xlpfnHj9k/edit#heading=h.ttegrcfy18ck)

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/DENG-4335)

### CODE REVIEW

**Comment by jaredsnyder** [comment]
Did some validation comparing to the google sheet which can be found [here](https://colab.research.google.com/drive/1qBmEHYvk7N5ZFZeRh0l4v4dY86Ajpik1#scrollTo=ei9Y4bvO34gF).  We're off by <4%.

File: sql/moz-fx-data-shared-prod/ads_derived/desktop_tiles_forecast_inputs_v1/metadata.yaml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  Geo markets as of 2024/07/16: 'US','DE','FR','AU','CA','IT','ES','MX','BR','IN','GB','JP'.
+owners:
+  - sbetancourt@mozilla.com
+  - cmorales@mozilla.com
+labels:
+  incremental: true
+  schedule: monthly
+  change_controlled: true
+scheduling:
+  dag_name: bqetl_ads
```
Comment (curtismorales) [inline sql/moz-fx-data-shared-prod/ads_derived/desktop_tiles_forecast_inputs_v1/metadata.yaml:15]: If this is only running monthly, it needs to be a separate DAG
```suggestion
  dag_name: bqetl_ads_monthly
```

**Review by curtismorales — COMMENTED** [review]
This looks good, a couple real minor notes. The one thing is that you'll need to make a new DAG for this table, since schedules are ultimately tied to the DAG and not the table (if it were in `bqetl_ads` it would run every day)

You want to add something like this to the `dags.yaml` file in the base directory of this repo
```
bqetl_ads_monthly:
  default_args:
    depends_on_past: false
    email:
      - telemetry-alerts@mozilla.com
      - cmorales@mozilla.com
      - sbetancourt@mozilla.com
    email_on_failure: true
    email_on_retry: true
    end_date: null
    owner: cmorales@mozilla.com
    retries: 2
    retry_delay: 30m
    start_date: '2024-08-01'
  description: Ads-related tables that run monthly
  repo: bigquery-etl
  schedule_interval: monthly
  tags:
    - impact/tier_1
    - repo/bigquery-etl
```

Oh and because of depencies, this might actually get a little more complicated. Actually if this is a pretty light query (I think it should be?) maybe it's best to just run this daily, in which case we'll need to adjust the metadata a little. Let's talk next week

**Comment by curtismorales** [comment]
Oh, and I'll look at the build on Monday as well.

File: sql/moz-fx-data-shared-prod/ads_derived/desktop_tiles_forecast_inputs_v1/query.sql [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+LEFT JOIN
+  cs_impressions AS c
+  ON c.country_code = n.country_code
+  AND c.submission_month = n.submission_month
+LEFT JOIN
+  users_table u
+  ON u.country_code = n.country_code
+  AND u.submission_month = n.submission_month
+WHERE
+  n.submission_month < DATE_TRUNC(@submission_date, MONTH)
```
[the remark this answers is not in the document]
Comment (jaredsnyder) [reply sql/moz-fx-data-shared-prod/ads_derived/desktop_tiles_forecast_inputs_v1/query.sql:107]: In the case where the "backfill" happens, this prevents there from being a little nub of data from the current month.  It's in the non-backfill case so really I should add it with the backfill case for clarity.  Fixed in most recent commit

File: sql/moz-fx-data-shared-prod/ads_derived/desktop_tiles_forecast_inputs_v1/schema.yaml [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  - mode: NULLABLE
+    name: visits_total_inventory_3
+    type: INTEGER
+    description: Count of sponsored tile inventory for positions 1 to 3 for submission_month. Calculated from latest Newtab data.
+  - mode: NULLABLE
+    name: visits_total_fill_rate_3
+    type: FLOAT
+    description: |-
+      Float of calculated fill rate (positions 1 to 3). Impressions / Inventory.
+      Calculated from latest Newtab and contextual-services data.
```
[the remark this answers is not in the document]
Comment (jaredsnyder) [reply sql/moz-fx-data-shared-prod/ads_derived/desktop_tiles_forecast_inputs_v1/schema.yaml:49]: Yeah agree it is misleading as-is.  I guess we could potentially add a new space for ads (lol) so maybe `1to3`?

[the remark this answers is not in the document]
Comment (jaredsnyder) [reply sql/moz-fx-data-shared-prod/ads_derived/desktop_tiles_forecast_inputs_v1/schema.yaml:49]: fixed in latest commit

File: sql/moz-fx-data-shared-prod/ads_derived/desktop_tiles_forecast_inputs_v1/schema.yaml [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    type: INTEGER
+    description: Count of sponsored tile inventory for positions 1 and 2 for submission_month. Calculated from latest Newtab data.
+  - mode: NULLABLE
+    name: visits_total_fill_rate_1and2
+    type: FLOAT
+    description: |-
+      Float of sponsored tile fill rate for positions 1 and 2 for submission_month. Impressions / Inventory.
+      Calculated from Newtab and contextual-services data.
+  - mode: NULLABLE
+    name: visits_total_inventory_3
```
Comment (curtismorales) [inline sql/moz-fx-data-shared-prod/ads_derived/desktop_tiles_forecast_inputs_v1/schema.yaml:41]:
```suggestion
    name: visits_total_inventory_1to3
```

File: sql/moz-fx-data-shared-prod/ads_derived/desktop_tiles_forecast_inputs_v1/schema.yaml [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    type: INTEGER
+    description: Count of sponsored tile inventory for positions 1 to 3 for submission_month. Calculated from latest Newtab data.
+  - mode: NULLABLE
+    name: visits_total_fill_rate_3
```
Comment (curtismorales) [inline sql/moz-fx-data-shared-prod/ads_derived/desktop_tiles_forecast_inputs_v1/schema.yaml:45]:
```suggestion
    name: visits_total_fill_rate_1to3
```

File: sql/moz-fx-data-shared-prod/ads_derived/desktop_tiles_forecast_inputs_v1/query.sql [hunk 6]
```diff
@@ -0,0 +1,108 @@
+WITH cs_impressions AS (
+  SELECT
+    country AS country_code,
+    DATE_TRUNC(submission_date, MONTH) AS submission_month,
+    SUM(CASE WHEN position <= 2 THEN event_count END) AS sponsored_impressions_1and2,
```
Comment (curtismorales) [inline sql/moz-fx-data-shared-prod/ads_derived/desktop_tiles_forecast_inputs_v1/query.sql:5]: Just a style tweak

```suggestion
    SUM(IF(position <= 2, event_count, 0)) AS sponsored_impressions_1and2,
```

File: sql/moz-fx-data-shared-prod/ads_derived/desktop_tiles_forecast_inputs_v1/query.sql [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+  u.total_user_count AS user_count,
+  c.sponsored_impressions_1and2 AS impression_count_1and2,
+  c.sponsored_impressions_all,
+  n.newtab_visits AS visit_count,
+  n.newtab_clients AS clients,
+  n.visits_total_inventory_1and2 AS total_inventory_1and2,
+  ROUND(1.00 * c.sponsored_impressions_1and2 / n.visits_total_inventory_1and2, 3) AS fill_rate,
+  n.visits_total_inventory_1to3,
+  ROUND(
+    1.00 * c.sponsored_impressions_all / n.visits_total_inventory_1to3,
```
Comment (curtismorales) [inline sql/moz-fx-data-shared-prod/ads_derived/desktop_tiles_forecast_inputs_v1/query.sql:93]: Same question here

File: sql/moz-fx-data-shared-prod/ads_derived/desktop_tiles_forecast_inputs_v1/query.sql [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+SELECT
+  n.submission_month,
+  n.country_code AS country,
+  u.total_user_count AS user_count,
+  c.sponsored_impressions_1and2 AS impression_count_1and2,
+  c.sponsored_impressions_all,
+  n.newtab_visits AS visit_count,
+  n.newtab_clients AS clients,
+  n.visits_total_inventory_1and2 AS total_inventory_1and2,
+  ROUND(1.00 * c.sponsored_impressions_1and2 / n.visits_total_inventory_1and2, 3) AS fill_rate,
```
Comment (curtismorales) [inline sql/moz-fx-data-shared-prod/ads_derived/desktop_tiles_forecast_inputs_v1/query.sql:90]: Do you need this `1.00 *`? I'm not clear on what this is doing; the division should work without it.

Comment (curtismorales) [inline sql/moz-fx-data-shared-prod/ads_derived/desktop_tiles_forecast_inputs_v1/query.sql:90]: Do you need this `1.00 *`? I'm not clear on what this is doing; the division should work without it.

File: sql/moz-fx-data-shared-prod/ads_derived/desktop_tiles_forecast_inputs_v1/query.sql (hunk 7 above)
Comment (curtismorales) [inline sql/moz-fx-data-shared-prod/ads_derived/desktop_tiles_forecast_inputs_v1/query.sql:93]: Same question here

File: sql/moz-fx-data-shared-prod/ads_derived/desktop_tiles_forecast_inputs_v1/metadata.yaml [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+  Geo markets as of 2024/07/16: 'US','DE','FR','AU','CA','IT','ES','MX','BR','IN','GB','JP'.
+owners:
+  - sbetancourt@mozilla.com
+  - cmorales@mozilla.com
+labels:
+  incremental: true
+  schedule: daily
+  change_controlled: true
+scheduling:
+  dag_name: bqetl_ads
```
Comment (curtismorales) [inline sql/moz-fx-data-shared-prod/ads_derived/desktop_tiles_forecast_inputs_v1/metadata.yaml:15]: 👍🏼

File: sql/moz-fx-data-shared-prod/ads_derived/desktop_tiles_forecast_inputs_v1/metadata.yaml [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+bigquery:
+  time_partitioning:
+    field: submission_month
+    type: day
+    require_partition_filter: true
```
Comment (curtismorales) [inline sql/moz-fx-data-shared-prod/ads_derived/desktop_tiles_forecast_inputs_v1/metadata.yaml:20]: You could probably leave this out; this table will always be pretty small I assume.

```suggestion
```

**Review by curtismorales — APPROVED** [review]
This looks good to me! I added a few style suggestions, but this ought to work as is.

I'm still going to check out the airflow scheduling thing and once I figure that out I may open a PR to move this to a monthly dag, but this is totally fine.

File: sql/moz-fx-data-shared-prod/ads_derived/desktop_tiles_forecast_inputs_v1/schema.yaml [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+  - mode: NULLABLE
+    name: clients
+    type: INTEGER
+    description: Count of distinct newtab clients opening the Newtab page for submission_month. Calculated from Newtab.
+  - mode: NULLABLE
+    name: total_inventory_1and2
+    type: INTEGER
+    description: Count of sponsored tile inventory for positions 1 and 2 for submission_month. Calculated from latest Newtab data.
+  - mode: NULLABLE
+    name: visits_total_fill_rate_1and2
```
Comment (curtismorales) [inline sql/moz-fx-data-shared-prod/ads_derived/desktop_tiles_forecast_inputs_v1/schema.yaml:35]: This field doesn't appear in the query; did you mean `visits_total_fill_rate_1to3`?
📎

**Comment by curtismorales** [comment]
Correction -- almost ready to go, there are some edits to make the SQL and schema.yaml match to get CI to pass

File: sql/moz-fx-data-shared-prod/ads_derived/desktop_tiles_forecast_inputs_v1/query.sql (hunk 8 above)
Comment (jaredsnyder) [reply sql/moz-fx-data-shared-prod/ads_derived/desktop_tiles_forecast_inputs_v1/query.sql:90]: I'm not sure what it's doing, @sergiosonline ?

File: sql/moz-fx-data-shared-prod/ads_derived/desktop_tiles_forecast_inputs_v1/schema.yaml (hunk 11 above)
Comment (jaredsnyder) [reply sql/moz-fx-data-shared-prod/ads_derived/desktop_tiles_forecast_inputs_v1/schema.yaml]: Should be `fill_rate`, pushed a fix

File: sql/moz-fx-data-shared-prod/ads_derived/desktop_tiles_forecast_inputs_v1/query.sql (hunk 8 above)
Comment (sergiosonline) [reply sql/moz-fx-data-shared-prod/ads_derived/desktop_tiles_forecast_inputs_v1/query.sql:90]: @jaredsnyder : At the time I put this together I wasn't confident BQ handled division btw integers well (unlike postgres and others). Just tested it and it works without the `1.00 *`. You can take it off.

**Pull request merged**